corehq/apps/app_manager/fields.py

```diff
@@ -116,3 +139,373 @@ def get_app_sources(domain):
+                                    form['app_copy'] = {'id': app.get_id, 'name': app.name}
+                                except KeyError:
+                                    form['app_copy'] = {'id': app.copy_of, 'name': '?'}
+                            if app.is_deleted():
+                                form['app_deleted'] = {'id': app.get_id}
+                            try:
+                                app_forms = app.get_xmlns_map()[form['xmlns']]
+                            except AttributeError:
+                                # it's a remote app
+                                app_forms = None
+                                form['has_app'] = True
+                            if app_forms:
+                                app_form = app_forms[0]
+                                if app_form.doc_type == 'UserRegistrationForm':
+                                    form['is_user_registration'] = True
+                                else:
+                                    app_module = app_form.get_module()
+                                    form['module'] = app_module
+                                    form['form'] = app_form
+                                form['show_xmlns'] = False
+
+                            if not form.get('app_copy') and not form.get('app_deleted'):
+                                form['no_suggestions'] = True
+                    if app:
+                        form['app'] = {'id': app.get_id, 'name': app.name, 'langs': app.langs}
+                else:
+                    form['possibilities'] = possibilities[form['xmlns']]
+                    if form['possibilities']:
+                        form['duplicate'] = True
+                    else:
+                        form['no_suggestions'] = True
+        return unknown_forms
+
+    @staticmethod
+    def _sorkey_form(form):
+        app_id = form['app']['id']
+        if form.get('has_app', False):
+            order = 0 if not form.get('app_deleted') else 1
+            app_name = form['app']['name']
+            module = form.get('module')
+            if module:
+                # module is sometimes wrapped json, sometimes a dict!
+                module_id = module['id'] if 'id' in module else module.id
+            else:
+                module_id = -1 if form.get('is_user_registration') else 1000
+            app_form = form.get('form')
+            if app_form:
+                # app_form is sometimes wrapped json, sometimes a dict!
+                form_id = app_form['id'] if 'id' in app_form else app_form.id
+            else:
+                form_id = -1
+            return (order, app_name, app_id, module_id, form_id)
+        else:
+            form_xmlns = form['xmlns']
+            return (2, form_xmlns, app_id)
+
+    @property
+    @memoized
+    def _all_forms(self):
+        forms = []
+        unknown_forms = []
+        for f in get_exports_by_form(self.domain):
```

Comment:
see moving code around. ALL of this functionality was asked for after the initial round of QA and then UAT feedback when we realized we had missed this functionality from the new ui, and it existed in reports/standard/exports.py which before lived here: https://github.com/dimagi/commcare-hq/blame/7b336da54786212c1cf856c6f0eddd3c171bed70/corehq/apps/reports/standard.py#L983. The goal of this feature was to reorganize the workflow and do more front end work to make the UX better, a lot of the backend stuff was code moved around, which I cleaned up a little, but given time pressures decided to table a more thorough cleanup for a point in the future since all of this was already a mess. I at least knew that what was there worked and covered all the edge cases